jepsen/src/jepsen/checker.clj

```diff
@@ -382,34 +380,51 @@
              upper              0             ; Upper bound on counter value
              pending-reads      {}            ; Process ID -> [lower read-val]
              reads              []]           ; Completed [lower val upper]s
-          (if (nil? history)
-            ; We're done here
-            (let [errors (remove (partial apply <=) reads)]
-              {:valid?             (empty? errors)
-               :reads              reads
-               :errors             errors})
-            ; But wait, there's more
-            (let [op      (first history)
-                  history (next history)]
-              (case [(:type op) (:f op)]
-                [:invoke :read]
+        (if (nil? history)
+          ; We're done here
+          (let [errors (remove (partial apply <=) reads)]
+            {:valid?             (empty? errors)
+             :reads              reads
+             :errors             errors})
+          ; But wait, there's more
+          (let [op      (first history)
+                history (next history)]
+            (case [(:type op) (:f op)]
+              [:invoke :read]
+              (recur history lower upper
+                     (assoc pending-reads (:process op) [[lower upper]])
+                     reads)
+
+              [:ok :read]
+              (let [read-ranges (get pending-reads (:process op))
+                    v (:value op)
+                    [l' u'] (first read-ranges)
+                    read (or (some (fn [[l u]] (when (<= l v u) [l v u])) read-ranges)
```

Comment:
I'm slightly confused here. Can you pls explain what you are doing?

read is assigned to the first matching [l u] which satisfies (<= l v u).
Why do that? Doesn't (:process op) tell us the op that the value is for?

Reply:
I've added more comments to clarify how this checker works. As for `(:process op)` - this is ID of the process, not the value. We are trying to find a range within possible ranges which contains the value `v` we've received from read op. If we can't find it - this is an inconsistency and we just save first range and `v` into `reads`, so later when we check for errors we see that `l <= v <= r` is not satisfied and report an error.